MNT Support and test Python 3.9

terencehonles opened the pull request:
#### What does this implement/fix? Explain your changes.

Updates Python 3.8 to Python 3.9 in latest tests, and updates trove classifiers to mention Python 3.9 support. I'm not sure what else is needed to build the Python 3.9 wheels since I didn't see anything obvious in this repo that should be changed.

terencehonles committed:
MNT Support and test Python 3.9

The following were used as references when updating pyproject.toml:

- https://github.com/scipy/scipy/issues/12940
- https://github.com/scikit-image/scikit-image/pull/4920
- https://github.com/scikit-image/scikit-image/pull/4920#discussion_r471467159

gsheni commented on build_tools/azure/install.sh:
```diff
@@ -84,6 +84,13 @@ elif [[ "$DISTRIB" == "conda-pip-latest" ]]; then
+    # NOTE(honles): scikit-image does not include pyproject.toml in its
+    #               current release and these are what it needs to build.
+    #               This can be removed when
+    #               https://github.com/scikit-image/scikit-image/pull/5016
```
That MR has been merged: https://github.com/scikit-image/scikit-image/pull/5016

terencehonles replied:
Yes, but it has not been released yet. I will create a separate PR which can be merged after that is released to pypi/conda.

terencehonles replied:
I'm also trying to get Python 3.9 wheels built too, so hopefully that will all come together in the next release

See: https://github.com/scikit-image/scikit-image/pull/5052

ogrisel replied:
I would rather not build scikit-image from source for now. Maybe we can only install scikit-image for the builds that are for Python `PYTHON_VERSION < "3.9"` for now instead.

It's an optional dependency of scikit-learn, so it should still work.

terencehonles replied:
So would you prefer I change this to: `python -m pip install --only-binary :all: sckit-image || true`? So it only installs binary packages and only if they exists? I'd have to see if anything else has to be altered, but I'd prefer not to add config that has to be updated manually and just install it automatically if it can.

ogrisel commented on pyproject.toml:
```diff
@@ -4,12 +4,12 @@ requires = [
+    "numpy==1.16.5; python_version=='3.7'",
+    "numpy==1.17.3; python_version=='3.8'",
+    # do not pin numpy on future versions of python to avoid incompatible numpy and python versions
+    "numpy; python_version>='3.9'",
```
We need to pin the oldest version of numpy that supports python_version == '3.9' because it will define the ABI compatibility expectations for the scikit-learn wheel.

terencehonles replied:
I'm not sure I follow this comment. This is the same as scipy, and I figured that would be acceptable. I'm not sure if you can explain your thinking a little more.

I *can* change it to 1.19.3 if needed, since that is the first version that has released wheels for Python 3.9. I just figured it might cut down on the maintenance burden and I assumed that was why scipy set that as their default.

ogrisel replied:
We use the version of numpy pyproject.toml to build the wheels in a github actions workflow. scikit-learn has Cython extension that need numpy at build time and the generated wheel need to be built with the oldest supported version of numpy to make sure that our build do not depend on recently changes of the numpy ABI.

For instance the future scikit-learn 0.24 wheel for python 3.8 needs to be build with numpy 1.17.3 to make  sure that users who have such old versions of numpy can install a binary package of scikit-learn that is binary compatible.

ogrisel replied:
The wheel building workflow is here:

https://github.com/scikit-learn/scikit-learn/blob/master/.github/workflows/wheels.yml

terencehonles replied:
@ogrisel yeah, I gathered as much as you explained. Not sure why scipy chose to not pin numpy the same way, but I'm fine with setting this to 1.19.3, and thanks for the pointer to the new workflow. cibuildwheels looks nice :slightly_smiling_face: 

@alfaro96 yeah I got that :wink: , but thanks for making it explicit

ogrisel reviewed (changes requested):
Thanks for the PR but we need to pin the version of numpy in the build dependency:

ogrisel commented:
/cc @alfaro96 since it's related to the new wheel builder infrastructure developed in  #17921.

alfaro96 commented on pyproject.toml:
```diff
@@ -4,12 +4,12 @@ requires = [
-    "numpy==1.16.0; python_version=='3.6' and platform_system=='AIX'",
-    "numpy==1.16.0; python_version=='3.7' and platform_system=='AIX'",
-    "numpy==1.17.3; python_version>='3.8' and platform_system=='AIX'",
+    # We follow scipy for much of these pinnings
+    # https://github.com/scipy/scipy/blob/master/pyproject.toml
+    "numpy==1.16.5; python_version=='3.6'",
+    "numpy==1.16.5; python_version=='3.7'",
+    "numpy==1.17.3; python_version=='3.8'",
+    # do not pin numpy on future versions of python to avoid incompatible numpy and python versions
+    "numpy; python_version>='3.9'",
```
According to @ogrisel, I would suggest the following changes:

```suggestion
    "numpy==1.16.5; python_version=='3.6'",
    "numpy==1.16.5; python_version=='3.7'",
    "numpy==1.17.3; python_version=='3.8'",
    "numpy==1.19.3; python_version>='3.9'",
```

The wheel builder uses the `pyproject.toml` file to know the build dependencies. Because we do not want that our build depends on any recent changes of the NumPy C-API, we must build the wheels with the oldest supported version of `NumPy`. By this way, we ensure that even users with old versions of `NumPy` can install the binary package of `scikit-learn` that is binary compatible, since we have previously ensured it.

terencehonles replied:
I restored the original pyproject.toml and just updated this for python 3.9, I sorted the requirements by Python version to make it more clear what was required. I figured it might make more sense to leave it as it was and only drop when Python 3.6 is dropped because there's more machinery that I was not aware of.

alfaro96 reviewed (changes requested):
Hey @terencehonles,

Thank you for your PR!

I have a few of requested changes:

* I think that we may update the Python version used in `circleci`:

```shell
grep -rnw ".circleci" -e "python"
.circleci/config.yml:6:      - image: circleci/python:3.7.3-stretch
.circleci/config.yml:47:      - image: circleci/python:3.7.3-stretch
.circleci/config.yml:93:      - image: circleci/python:3.6
.circleci/config.yml:122:      - image: circleci/python:3.6
```

* If we decide to update the `NumPy` version for Python 3.6 to `1.16.5`, we should update the `NUMPY_MIN_VERSION` global variable in the `sklearn/_min_dependencies.py` file.

terencehonles commented on .github/workflows/check-manifest.yml:
```diff
@@ -11,7 +11,7 @@ jobs:
       - uses: actions/setup-python@v2
         with:
-          python-version: '3.8'
+          python-version: '3.9'
```
Are you suggesting I remove this? I believe it might make sense to keep this explicit since this is a C extension that may need to be pinned to a different version than what is the default. If you'd like me to remove it I can.

terencehonles commented on azure-pipelines.yml:
```diff
@@ -15,7 +15,7 @@ jobs:
     - task: UsePythonVersion@0
       inputs:
-        versionSpec: '3.8'
+        versionSpec: '3.9'
```
Same comment here, though I believe Azure Pipelines may not have switched the default (I know this was discussed earlier on). I also remember coming across https://github.com/microsoft/azure-pipelines-tasks/issues/13708 but it looks like there has been no response there.

terencehonles commented:
> * I think that we may update the Python version used in `circleci`:

I'm not sure if we can update `doc-min-dependencies` I bumped it to use the Python 3.9 circle CI image, and I specified Python 3.9 as the Python version. It fails during conda install, but if I'm not mistaken that job looks like it's computing the min requirements and it makes sense to be running it with the lowest Python version. Does that seem correct?

ogrisel commented:
> It fails during conda install, but if I'm not mistaken that job looks like it's computing the min requirements and it makes sense to be running it with the lowest Python version. Does that seem correct?

Yes Python is considered a dependency of scikit-learn as numpy, scipy, etc. So the min_dependency build of circle CI should use the oldest supported version of Python which is Python 3.6 at the moment.

ogrisel commented:
> Not sure why scipy chose to not pin numpy the same way, but I'm fine with setting this to 1.19.3.

If scipy pins to an older version of numpy for Python 3.9, we can do so, but we need to wait for this Python 3.9 wheel for numpy to be uploaded to that version on PyPI, otherwise we won't be able to build our own wheels.

ogrisel commented:
@terencehonles To keep this PR focused and merge it quickly to prepare the scikit-learn 0.24 release, could you please revert the changes made on the circle ci configuration? They are not required to support and test Python 3.9.

We can always bump up the python versions used in circle ci in another PR but this requires more in dept testing and now is not the time to risk making the doc building infrastructure unstable ;)

Same for PyPy which is hard to test at the moment because of #18624.

alfaro96 commented:
@terencehonles Do you mind updating the `pyproject.toml` file according to #https://github.com/scikit-learn/scikit-learn/pull/18782#issuecomment-723464240?

terencehonles commented:
> > Not sure why scipy chose to not pin numpy the same way, but I'm fine with setting this to 1.19.3.
> 
> If scipy pins to an older version of numpy for Python 3.9, we can do so, but we need to wait for this Python 3.9 wheel for numpy to be uploaded to that version on PyPI, otherwise we won't be able to build our own wheels.

@ogrisel I'm aware of that, I mean that they don't constrain future releases to the min version needed to build. It's not really important to discuss since it seems like there are more specific requirements for this package.

> @terencehonles To keep this PR focused and merge it quickly to prepare the scikit-learn 0.24 release, could you please revert the changes made on the circle ci configuration? They are not required to support and test Python 3.9.
> 
> We can always bump up the python versions used in circle ci in another PR but this requires more in dept testing and now is not the time to risk making the doc building infrastructure unstable ;)
> 
> Same for PyPy which is hard to test at the moment because of #18624.

@ogrisel this PR really isn't that big and now two of you are asking for conflicting things. I changed it because @alfaro96 asked, but also because the change does need to be tested. The CI changes are not failing so I'm not sure what's your reason to not make the change now.



> @terencehonles Do you mind updating the `pyproject.toml` file according to [#18782 (comment)](https://github.com/scikit-learn/scikit-learn/pull/18782#issuecomment-723464240)?

@alfaro96 done, I added some comments to explain when the constraints can be removed as Python versions are expired.

ogrisel commented:
> The CI changes are not failing so I'm not sure what's your reason to not make the change now.

The full doc build is very expensive and not triggered by your PR. The PyPy build is only triggered by a scheduled task it does not run on PRs.

See: https://scikit-learn.org/dev/developers/contributing.html#continuous-integration-ci

terencehonles commented:
> > The CI changes are not failing so I'm not sure what's your reason to not make the change now.
> 
> The full doc build is very expensive and not triggered by your PR. The PyPy build is only triggered by a scheduled task it does not run on PRs.
> 
> See: https://scikit-learn.org/dev/developers/contributing.html#continuous-integration-ci

It really looks like that only applies to pypy which I had thought I hadn't included, but it looks like I had left it in. Since I was going to be pushing I reverted any changes to Circle CI at your request.

ogrisel reviewed (approved):
LGTM!

ogrisel committed:
Trigger wheel builder [cd build]

ogrisel commented:
I just pushed a commit to check that the new pyproject.toml does not break the wheel builder workflow.

alfaro96 commented on pyproject.toml:
```diff
@@ -4,12 +4,27 @@ requires = [
-    "numpy==1.16.0; python_version=='3.6' and platform_system=='AIX'",
-    "numpy==1.16.0; python_version=='3.7' and platform_system=='AIX'",
-    "numpy==1.17.3; python_version>='3.8' and platform_system=='AIX'",
+
+    # PyPy needs numpy >= 1.14.0
+    # platform_python_implementation!='CPython' not needed >= Python 3.7 which is numpy 1.14.5
+    "numpy==1.13.3; python_version=='3.6' and platform_machine!='aarch64' and platform_system!='AIX' and platform_python_implementation=='CPython'",
+    "numpy==1.14.0; python_version=='3.6' and platform_machine!='aarch64' and platform_system!='AIX' and platform_python_implementation!='CPython'",
+
+    # AIX needs numpy >= 1.16.0
+    # platform_system!='AIX' not needed >= Python 3.8 which is numpy 1.17.3
+    "numpy==1.16.0; python_version=='3.6' and platform_machine!='aarch64' and platform_system=='AIX'",
+    "numpy==1.16.0; python_version=='3.7' and platform_machine!='aarch64' and platform_system=='AIX'",
+
+    # ARM needs numpy >= 1.19.0
+    # platform_machine!='aarch64' not needed >= Python 3.9 which is numpy 1.19.3
+    "numpy==1.19.0; python_version=='3.6' and platform_machine=='aarch64'",
+    "numpy==1.19.0; python_version=='3.7' and platform_machine=='aarch64'",
+    "numpy==1.19.0; python_version=='3.8' and platform_machine=='aarch64'",
+
+    # default numpy requirements
+    "numpy==1.14.5; python_version=='3.7' and platform_machine!='aarch64' and platform_system!='AIX'",
+    "numpy==1.17.3; python_version=='3.8' and platform_machine!='aarch64'",
+    "numpy==1.19.3; python_version=='3.9'",
+
```
What about using [`oldest-supported-numpy`](https://pypi.org/project/oldest-supported-numpy/)?

It would simplify the `pyproject.toml` file quite a lot.

WDYT @thomasjpfan and @ogrisel?

terencehonles replied:
That seems like a pretty interesting idea to me, but we'd need to upstream all the changes that were made here (existing pypy constraints and added ARM constraints)

alfaro96 replied:
Let us accept this solution for now then and keep us track of `oldest-supported-numpy` to see if we could use it in the future.

ogrisel replied:
oldest-supported-numpy (implicitly by numpy developers) is not necessarily the same as oldest-supported-numpy by scikit-learn developers.

I think we are rather conservative, so the two might match, but we would have to discuss that with a dedicated issue.

Let's not block this PR further but feel free to open an issue on this.

terencehonles replied:
Do you know where the pypy version constraints come from? Looking at PyPI I can see the ARM wheels so I can create a PR for that, but I don't know if I should upstream the PyPy requirement of `1.14.0`

ogrisel replied:
Not sure either.

alfaro96 replied:
See https://github.com/scikit-learn/scikit-learn/pull/11010#issuecomment-396397514. We had 7 failures with `numpy==1.13.3` and none with `numpy==1.14.0` for `PyPy`.

I do not think that we could upstream such dependencies since as @ogrisel pointed out:  "`oldest-supported-numpy` is not necessarily the same as `oldest-supported-numpy` by `scikit-learn` developers."

Sorry for the noise!

alfaro96 reviewed (commented):
Thank you @terencehonles for your PR.

I would like to know if it would be possible to use `oldest-supported-numpy` to simplify the `pyproject.toml` file.

terencehonles commented on build_tools/azure/install.sh:
```diff
@@ -84,7 +84,10 @@ elif [[ "$DISTRIB" == "conda-pip-latest" ]]; then
 
-    python -m pip install pandas matplotlib pyamg scikit-image
+    # optionally install scikit-image
+    python -m pip install --only-binary :all: sckit-image || true
```
thanks :confused:

terencehonles commented on build_tools/azure/install.sh:
```diff
@@ -84,7 +84,10 @@ elif [[ "$DISTRIB" == "conda-pip-latest" ]]; then
     python -m pip install -U pip
 
-    python -m pip install pandas matplotlib pyamg scikit-image
+    # optionally install scikit-image
```
Do we actually want this removed? It will install scikit-image whenever it can, and when the Python 3.9 wheels are released they will be installed, and when Python 3.10 is released and there is no wheels updating this file will not require adding this back.

alfaro96 replied:
If we always want to install `scikit-image` from binaries (not sure about this), we should specify why.

Maybe something like:

```suggestion
    # Do not build scikit-image from source because it is an optional dependency
```

is clearer.

terencehonles committed:
Fix typo when optionally installing scikit-image

Co-authored-by: Juan Carlos Alfaro Jiménez <JuanCarlos.Alfaro@uclm.es>

alfaro96 reviewed (approved):
LGTM.

ogrisel committed:
Update build_tools/azure/install.sh [ci skip]

Co-authored-by: Juan Carlos Alfaro Jiménez <JuanCarlos.Alfaro@uclm.es>

adrinjalali reviewed (approved):
LGTM, and pretty small at this point.

terencehonles commented:
> Let us accept this solution and keep us track of `oldest-supported-numpy` to see if we could use it in the future.

I've opened a PR to upstream the aarch constraints https://github.com/scipy/oldest-supported-numpy/pull/13, but if the PyPy change doesn't get upstreamed you can probably use the `oldest-supported-numpy` package when dropping Python 3.6

ogrisel commented:
Thanks all!
